js/src/SearchBar.js

```diff
@@ -1,158 +1,171 @@
 
-var SearchBar = utils.make_class();
-// instance methods
-SearchBar.prototype = {
-    init: init,
-    is_visible: is_visible,
-    toggle: toggle,
-    update: update,
-    next: next,
-    previous: previous
-};
-module.exports = SearchBar;
+const _ = require('underscore')
 
-function init(sel, search_index, map) {
-    var container = sel.attr('class', 'search-container')
-            .style('display', 'none');
-    this.input = container.append('input')
-        .attr('class', 'search-bar');
-    var group = container.append('div').attr('class', 'btn-group btn-group-sm');
-    group.append('button')
-        .attr('class', 'btn btn-default')
-        .on('click', this.previous.bind(this))
-        .append('span').attr('class', 'glyphicon glyphicon-chevron-left');
-    group.append('button')
-        .attr('class', 'btn btn-default')
-        .on('click', this.next.bind(this))
-        .append('span').attr('class', 'glyphicon glyphicon-chevron-right');
-    this.counter = container.append('div')
-        .attr('class', 'search-counter');
-    container.append('button')
-        .attr('class', 'btn btn-sm btn-default close-button')
-        .on('click', function() {
-            this.toggle(false);
-        }.bind(this))
-        .append('span').attr('class',  'glyphicon glyphicon-remove');
-
-    this.callback_manager = new CallbackManager();
+class SearchBar extends Component {
+  constructor (props) {
+    super(props)
+    this.state = {
+      visible: props.visible,
+      current: 1,
+      searchItem: null,
+      counter: ''
+    }
+  }
 
-    this.selection = container;
-    this.map = map;
-    this.search_index = search_index;
+  componentWillReceiveProps (nextProps) {
+    this.props.map.key_manager.add_escape_listener(() => this.close(), true)
+    this.setState({
+      ...nextProps,
+      current: 1,
+      results: null,
+      searchItem: null,
+      counter: '',
+      clearNext: this.props.map.key_manager.add_key_listener(
+        ['enter', 'ctrl+g'], () => this.next(), false),
+      clearPrevious: this.props.map.key_manager.add_key_listener(
+        ['shift+enter', 'shift+ctrl+g'], () => this.previous(), false)
+    })
+  }
 
-    this.current = 1;
-    this.results = null;
+  componentDidUpdate () {
+    this.inputRef.focus()
+  }
 
-    var on_input_fn = function(input) {
-        this.current = 1;
-        this.results = _drop_duplicates(this.search_index.find(input.value));
-        this.update();
-    }.bind(this, this.input.node());
-    this.input.on('input', utils.debounce(on_input_fn, 200));
-}
-
-var comp_keys = {
-    metabolite: ['m', 'node_id'],
-    reaction: ['r', 'reaction_id'],
-    text_label: ['t', 'text_label_id']
-};
-function _drop_duplicates(results) {
-    return _.uniq(results, function(item) {
-        // make a string for fast comparison
-        var t = comp_keys[item.type];
-        return t[0] + item[t[1]];
-    });
-}
+  handleInput (value) {
+    if (this.state.visible) {
```

Comment:
There's no else condition for this if, an early return would probably make this piece of code more readable: 
```
handleInput(value) {
if (!this.state.visible) {return;}
const results = // ...
```

Reply:
Done